File: crates/runtime/engine/src/executor/hooks.rs
//! Scheduler-level execution hooks.
//!
//! These hooks complement dispatcher middleware. Middleware wraps the actual
//! operation handler, while execution hooks observe scheduler lifecycle events
//! such as ready, start, finish, and graph completion.

use std::sync::Arc;

use apxm_core::types::{AISOperationType, NodeId, OpStatus};

/// Observer interface for runtime execution lifecycle events.
pub trait ExecutionHook: Send + Sync {
    /// Stable hook name for diagnostics.
    fn name(&self) -> &str {
        "execution-hook"
    }

    /// Called when a graph starts running.
    fn on_graph_started(&self, _event: &GraphStartedEvent) {}

    /// Called when a node becomes ready and is admitted to the scheduler queue.
    fn on_node_ready(&self, _event: &NodeReadyEvent) {}

    /// Called when a worker starts executing a ready node.
    fn on_node_started(&self, _event: &NodeStartedEvent) {}

    /// Called when a node reaches a terminal scheduler state.
    fn on_node_finished(&self, _event: &NodeFinishedEvent) {}

    /// Called when a graph finishes or fails.
    fn on_graph_finished(&self, _event: &GraphFinishedEvent) {}
}

/// Immutable hook set for one graph execution.
#[derive(Clone, Default)]
pub struct ExecutionHookContext {
    execution_id: String,
    graph_id: String,
    hooks: Vec<Arc<dyn ExecutionHook>>,
}

impl ExecutionHookContext {
    pub fn new(
        execution_id: impl Into<String>,
        graph_id: impl Into<String>,
        hooks: Vec<Arc<dyn ExecutionHook>>,
    ) -> Self {
        Self {
            execution_id: execution_id.into(),
            graph_id: graph_id.into(),
            hooks,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.hooks.is_empty()
    }

    pub fn execution_id(&self) -> &str {
        &self.execution_id
    }

    pub fn graph_id(&self) -> &str {
        &self.graph_id
    }

    pub fn emit_graph_started(&self, node_count: usize) {
        if self.is_empty() {
            return;
        }
        let event = GraphStartedEvent {
            execution_id: self.execution_id.clone(),
            graph_id: self.graph_id.clone(),
            node_count,
        };
        for hook in &self.hooks {
            hook.on_graph_started(&event);
        }
    }

    pub fn emit_node_ready(&self, event: NodeReadyEvent) {
        if self.is_empty() {
            return;
        }
        for hook in &self.hooks {
            hook.on_node_ready(&event);
        }
    }

    pub fn emit_node_started(&self, event: NodeStartedEvent) {
        if self.is_empty() {
            return;
        }
        for hook in &self.hooks {
            hook.on_node_started(&event);
        }
    }

    pub fn emit_node_finished(&self, event: NodeFinishedEvent) {
        if self.is_empty() {
            return;
        }
        for hook in &self.hooks {
            hook.on_node_finished(&event);
        }
    }

    pub fn emit_graph_finished(
        &self,
        executed_nodes: usize,
        failed_nodes: usize,
        duration_ms: u128,
        success: bool,
    ) {
        if self.is_empty() {
            return;
        }
        let event = GraphFinishedEvent {
            execution_id: self.execution_id.clone(),
            graph_id: self.graph_id.clone(),
            executed_nodes,
            failed_nodes,
            duration_ms,
            success,
        };
        for hook in &self.hooks {
            hook.on_graph_finished(&event);
        }
    }
}

#[derive(Debug, Clone)]
pub struct GraphStartedEvent {
    pub execution_id: String,
    pub graph_id: String,
    pub node_count: usize,
}

#[derive(Debug, Clone)]
pub struct NodeReadyEvent {
    pub execution_id: String,
    pub graph_id: String,
    pub node_id: NodeId,
    pub op_type: AISOperationType,
    pub priority: String,
    pub ready_at_ms: u128,
}

#[derive(Debug, Clone)]
pub struct NodeStartedEvent {
    pub execution_id: String,
    pub graph_id: String,
    pub node_id: NodeId,
    pub op_type: AISOperationType,
    pub priority: String,
    pub worker_id: Option<usize>,
    pub ready_at_ms: Option<u128>,
    pub started_at_ms: u128,
    pub queue_wait_ms: Option<u128>,
}

#[derive(Debug, Clone)]
pub struct NodeFinishedEvent {
    pub execution_id: String,
    pub graph_id: String,
    pub node_id: NodeId,
    pub op_type: AISOperationType,
    pub status: OpStatus,
    pub attempts: u32,
    pub started_at_ms: Option<u128>,
    pub finished_at_ms: u128,
    pub duration_ms: Option<u128>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct GraphFinishedEvent {
    pub execution_id: String,
    pub graph_id: String,
    pub executed_nodes: usize,
    pub failed_nodes: usize,
    pub duration_ms: u128,
    pub success: bool,
}

// ===========================================================================
// Program-authored lifecycle hooks (the `@hook` mechanism).
//
// Distinct from the `ExecutionHook` observer trait above: these are
// program-authored bindings (from the artifact's hooks sidecar / `REGISTER_HOOK`
// op) that drive the SAME Python handler path as `@tool` (constitution #4). The
// registry lives on `ExecutionContext` (lifetime = the python tool bridge) and
// is inherited by child contexts. Drivers (the interceptor/middleware/async
// pre-step) consume it.
// ===========================================================================

/// Conversational lifecycle events a `@hook` can bind to.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HookEvent {
    SessionStart,
    PreTurn,
    PostTurn,
    PreAsk,
    PostAsk,
    PreCap,
    PostCap,
}

impl HookEvent {
    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "session_start" => Some(Self::SessionStart),
            "pre_turn" => Some(Self::PreTurn),
            "post_turn" => Some(Self::PostTurn),
            "pre_ask" => Some(Self::PreAsk),
            "post_ask" => Some(Self::PostAsk),
            "pre_cap" => Some(Self::PreCap),
            "post_cap" => Some(Self::PostCap),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::SessionStart => "session_start",
            Self::PreTurn => "pre_turn",
            Self::PostTurn => "post_turn",
            Self::PreAsk => "pre_ask",
            Self::PostAsk => "post_ask",
            Self::PreCap => "pre_cap",
            Self::PostCap => "post_cap",
        }
    }

    /// Pre-execution events are the only ones that may `gate` (allow/deny/edit).
    pub fn is_pre(&self) -> bool {
        matches!(
            self,
            Self::SessionStart | Self::PreTurn | Self::PreAsk | Self::PreCap
        )
    }
}

/// Whether a hook may only observe or may gate (control) the guarded action.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HookMode {
    Observe,
    Gate,
}

impl HookMode {
    pub fn parse(s: &str) -> Option<Self> {
        match s {
            "observe" => Some(Self::Observe),
            "gate" => Some(Self::Gate),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Observe => "observe",
            Self::Gate => "gate",
        }
    }
}

/// One author hook binding: a Python handler bound to a lifecycle event.
#[derive(Debug, Clone)]
pub struct HookBinding {
    /// Python handler id, dispatched via the shared `PythonHandlerBridge`.
    pub handler_id: String,
    pub event: HookEvent,
    /// Glob over tool/op name this hook applies to (`*` = all).
    pub match_glob: String,
    pub mode: HookMode,
}

impl HookBinding {
    /// Does this hook apply to a tool/op named `name`? Supports `*` wildcards.
    pub fn matches(&self, name: &str) -> bool {
        glob_match(&self.match_glob, name)
    }
}

/// Per-artifact registry of author hook bindings, resolved from `REGISTER_HOOK`
/// nodes at runtime (and/or the hooks sidecar at load). Threaded as
/// `Arc<HookRegistry>` on `ExecutionContext` (lifetime = the python tool bridge)
/// and inherited by child contexts, so registration by a `REGISTER_HOOK` node in
/// the entry flow is visible to later turn/tool nodes. Interior mutability lets
/// the shared (`Arc`) registry be populated through `&self`.
#[derive(Debug, Default)]
pub struct HookRegistry {
    bindings: std::sync::Mutex<Vec<HookBinding>>,
}

impl HookRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn is_empty(&self) -> bool {
        self.bindings
            .lock()
            .expect("hook registry poisoned")
            .is_empty()
    }

    pub fn len(&self) -> usize {
        self.bindings.lock().expect("hook registry poisoned").len()
    }

    pub fn register(&self, binding: HookBinding) {
        self.bindings
            .lock()
            .expect("hook registry poisoned")
            .push(binding);
    }

    /// All bindings for `event`, in registration order (owned clones so the
    /// lock is not held by the caller).
    pub fn for_event(&self, event: HookEvent) -> Vec<HookBinding> {
        self.bindings
            .lock()
            .expect("hook registry poisoned")
            .iter()
            .filter(|b| b.event == event)
            .cloned()
            .collect()
    }

    /// Bindings for `event` whose `match` glob applies to `name`.
    pub fn matching(&self, event: HookEvent, name: &str) -> Vec<HookBinding> {
        self.bindings
            .lock()
            .expect("hook registry poisoned")
            .iter()
            .filter(|b| b.event == event && b.matches(name))
            .cloned()
            .collect()
    }
}

/// Minimal glob match supporting `*` wildcards (anywhere in the pattern). An
/// empty pattern or `*` matches everything.
fn glob_match(pattern: &str, name: &str) -> bool {
    if pattern.is_empty() || pattern == "*" {
        return true;
    }
    if !pattern.contains('*') {
        return pattern == name;
    }
    // Split on `*` and match the literal segments in order, anchoring the first
    // and last segments to the start/end unless the pattern begins/ends with `*`.
    let segments: Vec<&str> = pattern.split('*').collect();
    let mut pos = 0usize;
    for (i, seg) in segments.iter().enumerate() {
        if seg.is_empty() {
            continue;
        }
        // `name.get(pos..)` returns None if `pos` is not a char boundary, so a
        // non-ASCII name can never panic on a byte slice (returns no-match).
        let rest = match name.get(pos..) {
            Some(r) => r,
            None => return false,
        };
        if i == 0 {
            if !rest.starts_with(seg) {
                return false;
            }
            pos += seg.len();
        } else if i == segments.len() - 1 {
            if !rest.ends_with(seg) {
                return false;
            }
        } else if let Some(found) = rest.find(seg) {
            pos += found + seg.len();
        } else {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod hook_registry_tests {
    use super::*;

    #[test]
    fn event_round_trips() {
        for s in [
            "session_start",
            "pre_turn",
            "post_turn",
            "pre_ask",
            "post_ask",
            "pre_cap",
            "post_cap",
        ] {
            assert_eq!(HookEvent::parse(s).unwrap().as_str(), s);
        }
        assert!(HookEvent::parse("nope").is_none());
    }

    #[test]
    fn only_pre_events_gate() {
        assert!(HookEvent::PreCap.is_pre());
        assert!(HookEvent::SessionStart.is_pre());
        assert!(!HookEvent::PostCap.is_pre());
        assert!(!HookEvent::PostAsk.is_pre());
    }

    #[test]
    fn glob_matching() {
        assert!(glob_match("*", "lookup"));
        assert!(glob_match("", "lookup"));
        assert!(glob_match("lookup", "lookup"));
        assert!(!glob_match("lookup", "search"));
        assert!(glob_match("look*", "lookup"));
        assert!(glob_match("*up", "lookup"));
        assert!(glob_match("l*p", "lookup"));
        assert!(!glob_match("x*y", "lookup"));
        // Non-ASCII names must not panic on byte slicing (m2).
        assert!(glob_match("*", "café_tool"));
        assert!(glob_match("café*", "café_tool"));
        assert!(!glob_match("a*z", "café"));
        assert!(glob_match("c*é", "café"));
    }

    /// The turn/ask lifecycle events the turn drivers consume (`pre_turn`,
    /// `post_turn`, `post_ask`) must be registrable and discoverable via
    /// `for_event` — i.e. not dead surface.
    #[test]
    fn lifecycle_turn_events_are_queryable() {
        let reg = HookRegistry::new();
        for ev in [HookEvent::PreTurn, HookEvent::PostTurn, HookEvent::PostAsk] {
            reg.register(HookBinding {
                handler_id: format!("h-{}", ev.as_str()),
                event: ev,
                match_glob: "*".into(),
                mode: HookMode::Observe,
            });
        }
        assert_eq!(reg.for_event(HookEvent::PreTurn).len(), 1);
        assert_eq!(reg.for_event(HookEvent::PostTurn).len(), 1);
        assert_eq!(reg.for_event(HookEvent::PostAsk).len(), 1);
        // post_turn carries the reply (per contract); event round-trips.
        assert_eq!(
            reg.for_event(HookEvent::PostTurn)[0].handler_id,
            "h-post_turn"
        );
    }

    #[test]
    fn registry_filters_by_event_and_match() {
        let reg = HookRegistry::new();
        reg.register(HookBinding {
            handler_id: "h1".into(),
            event: HookEvent::PreCap,
            match_glob: "lookup".into(),
            mode: HookMode::Gate,
        });
        reg.register(HookBinding {
            handler_id: "h2".into(),
            event: HookEvent::PreCap,
            match_glob: "*".into(),
            mode: HookMode::Observe,
        });
        reg.register(HookBinding {
            handler_id: "h3".into(),
            event: HookEvent::PostCap,
            match_glob: "*".into(),
            mode: HookMode::Observe,
        });
        assert_eq!(reg.len(), 3);
        let m = reg.matching(HookEvent::PreCap, "lookup");
        assert_eq!(m.len(), 2);
        let m2 = reg.matching(HookEvent::PreCap, "search");
        assert_eq!(m2.len(), 1);
        assert_eq!(m2[0].handler_id, "h2");
    }
}

File: crates/compiler/frontend/python/apxm/hooks.py
"""@hook decorator — bind a Python callable to a conversational lifecycle event.

A hook reuses the SAME invocation path as `@tool`: it produces a stable
`handler_id` and registers the callable in the shared tool registry, so the
runtime `PythonHandlerBridge` dispatches tools and hooks through one mechanism
(constitution #4 — one Python-handler mechanism). The frontend records each
hook as a `REGISTER_HOOK` op, so the binding travels inside canonical AIR.

This module owns conversational lifecycle hooks. The command-based hooks in
``config.HookConfig`` are a separate surface and are unaffected here.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable


class LifecycleEvent(str, Enum):
    SESSION_START = "session_start"
    PRE_TURN = "pre_turn"
    POST_TURN = "post_turn"
    PRE_ASK = "pre_ask"
    POST_ASK = "post_ask"
    PRE_CAP = "pre_cap"
    POST_CAP = "post_cap"


class HookMode(str, Enum):
    OBSERVE = "observe"
    GATE = "gate"


LIFECYCLE_EVENTS: frozenset[str] = frozenset(event.value for event in LifecycleEvent)
GATE_LIFECYCLE_EVENTS: frozenset[str] = frozenset(
    {
        LifecycleEvent.SESSION_START.value,
        LifecycleEvent.PRE_TURN.value,
        LifecycleEvent.PRE_ASK.value,
        LifecycleEvent.PRE_CAP.value,
    }
)
HOOK_MODES: frozenset[str] = frozenset(mode.value for mode in HookMode)


def normalize_lifecycle_event(value: LifecycleEvent | str) -> str:
    if isinstance(value, LifecycleEvent):
        return value.value
    if isinstance(value, str):
        return value
    raise TypeError("hook event must be a LifecycleEvent or string")


def normalize_hook_mode(value: HookMode | str) -> str:
    if isinstance(value, HookMode):
        return value.value
    if isinstance(value, str):
        return value
    raise TypeError("hook mode must be a HookMode or string")


def _make_handler_id(fn: Callable[..., Any]) -> str:
    """Stable handler id from module:qualname."""
    module = getattr(fn, "__module__", "__unknown__") or "__unknown__"
    qualname = getattr(fn, "__qualname__", fn.__name__)
    key = f"{module}:{qualname}"
    return f"sha256:{hashlib.sha256(key.encode('utf-8')).hexdigest()}"


@dataclass
class HookFn:
    """A Python callable bound to a lifecycle event.

    # Attributes
        fn: The original callable.
        event: One of ``LIFECYCLE_EVENTS``.
        match: Glob over tool/op name (default ``*``).
        mode: ``observe`` | ``gate``.
        handler_id: Stable content hash shared with the tool bridge.
        name: Hook name (defaults to the function name).
    """

    fn: Callable[..., Any]
    event: str
    match: str
    mode: str
    handler_id: str
    name: str
    metadata: dict[str, Any] = field(default_factory=dict)

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        return self.fn(*args, **kwargs)


def hook(
    *,
    on: LifecycleEvent | str,
    match: str = "*",
    mode: HookMode | str = HookMode.OBSERVE,
) -> Callable[[Callable[..., Any]], HookFn]:
    """Bind a Python callable to a conversational lifecycle event.

    Args:
        on: The lifecycle event (one of ``LIFECYCLE_EVENTS``).
        match: Glob over the tool/op name the hook applies to (default ``*``).
        mode: ``observe`` (default) or ``gate`` (allow/deny/edit; ``pre_*`` only).
    """
    event_value = normalize_lifecycle_event(on)
    mode_value = normalize_hook_mode(mode)
    if event_value not in LIFECYCLE_EVENTS:
        valid = ", ".join(sorted(LIFECYCLE_EVENTS))
        raise ValueError(f"@hook(on={event_value!r}) is not a valid event; expected one of: {valid}")
    if mode_value not in HOOK_MODES:
        valid = ", ".join(sorted(HOOK_MODES))
        raise ValueError(f"@hook(mode={mode_value!r}) invalid; expected one of: {valid}")
    if mode_value == HookMode.GATE.value and event_value not in GATE_LIFECYCLE_EVENTS:
        raise ValueError(
            f"@hook(on={event_value!r}, mode='gate') is invalid; gate mode is only "
            f"permitted on pre-execution events: {', '.join(sorted(GATE_LIFECYCLE_EVENTS))}"
        )

    def _wrap(fn: Callable[..., Any]) -> HookFn:
        # Reuse the shared tool registry so the bridge dispatches hooks and
        # tools through one path (constitution #4).
        from .tools import _TOOL_REGISTRY

        handler_id = _make_handler_id(fn)
        _TOOL_REGISTRY[handler_id] = fn
        return HookFn(
            fn=fn,
            event=event_value,
            match=match,
            mode=mode_value,
            handler_id=handler_id,
            name=getattr(fn, "__name__", "hook"),
        )

    return _wrap


def hook_descriptor(h: HookFn) -> dict[str, Any]:
    """Build a handler manifest descriptor for a hook."""
    import inspect

    module = getattr(h.fn, "__module__", "__unknown__") or "__unknown__"
    qualname = getattr(h.fn, "__qualname__", h.fn.__name__)
    descriptor: dict[str, Any] = {
        "handler_id": h.handler_id,
        "module": module,
        "qualname": qualname,
        "name": h.name,
        "event": h.event,
        "match": h.match,
        "mode": h.mode,
    }
    source_file = inspect.getsourcefile(h.fn)
    if source_file:
        descriptor["source_file"] = source_file
    return descriptor


__all__ = [
    "GATE_LIFECYCLE_EVENTS",
    "HookFn",
    "HookMode",
    "HOOK_MODES",
    "LifecycleEvent",
    "LIFECYCLE_EVENTS",
    "hook",
    "hook_descriptor",
    "normalize_hook_mode",
    "normalize_lifecycle_event",
]
